### ISSUE TITLE
Why JIT compilation Error?

### ISSUE THREAD

**Issue by BookmanHan** [arrayfire/arrayfire#1942]
I am sorry for always asking stupid questions, but I really work hard on it before asking.

I encounter a JIT issue as below:
```
what():  ArrayFire Exception (Internal error:998):
In function std::vector<char> cuda::compileToPTX(const char*, std::string)
In file src/backend/cuda/jit.cpp:264
NVRTC Error(6): NVRTC_ERROR_COMPILATION

In function void af::array::eval() const
In file src/api/cpp/array.cpp:956
```

My codes are a bit complex.
Simply, I implement a function, within which there are a lot of array operations then return an array.

Name the function as `af::array get();`
When I use `auto a = get();`, there is no error.
When I use `auto a = get() + get() + get();`, it's still normal.

But when I call this function more than there times, there comes up with the above exception.

What is the problem?
How should I fix it?

Thanks a lot.

**Comment by pavanky** [comment]
@BookmanHan This is harder to debug without seeing code. But if you cant share code, would you be willing to build arrayfire from source?

**Comment by umar456** [comment]
@pavanky it sounds like the function is failing when the JIT engine actually eval the expression created by the calls. I think it has to do with the NVRTC paths.

**Comment by pavanky** [comment]
@umar456 I know but I cant debug it without seeing what the error is. We should probably have a way to display verbose errors from JIT as a flag.

@BookmanHan You seem to be using an older version of arrayfire. There was a bug in JIT that @umar456 fixed for 3.5.1. Can you check if this is still an issue on your end?

**Comment by BookmanHan** [comment]
@pavanky 

I update to 3.5.1, but it does not fix the bugs.

Codes are https://github.com/BookmanHan/Bearlet
It's a calculus graph with ArrayFire, simply, as an automatic derivation toolset, like TensorFlow.

In Bearlet.cpp:
Each symbol object has two arrays, which are respectively the tensor value and its gradient.

Bugs stem from here:
```
Symbol* unit_1 = &unit(model, triple_user, triple_item, rate_1);
for(auto i=0; I<num_iter; ++i)
	unit_1 = &(*unit_1 + unit(model, triple_user, triple_item, rate_1));
	
Symbol* unit_5 = &unit(model, triple_user, triple_item, rate_5);
for(auto i=0; I<num_iter; ++i)
	unit_5 = &(*unit_5 + unit(model, triple_user, triple_item, rate_5));
```
When `num_iter = 9`, there is no error.
When `num_iter = 19`, there comes up with bugs, such below:
```
 what():  ArrayFire Exception (Internal error:998):
In function std::vector<char> cuda::compileToPTX(const char*, std::string)
In file src/backend/cuda/jit.cpp:291
NVRTC Error(6): NVRTC_ERROR_COMPILATION

In function void af::array::eval() const
In file src/api/cpp/array.cpp:964
```

I am very grateful for response.
What should I do next?

**Comment by pavanky** [comment]
@BookmanHan Thanks for the code I'll try to reproduce it. If you are trying to create a reverse mode autodiff, I implemented something here using arrayfire: https://github.com/arrayfire/arrayfire-ml

Examples can be seen here
- https://github.com/arrayfire/arrayfire-ml/blob/master/examples/autograd.cpp
- https://github.com/arrayfire/arrayfire-ml/blob/master/examples/xor.cpp

**Comment by BookmanHan** [comment]
@pavanky 
It is very nice of you to introduce arrayfire-ml project.
But, actually, I am implementing something new onto the auto-diff framework for research purpose.

I really hope you could fix this issue.
Because I conjecture the issue may be incurred by overloaded computations.
Maybe, even if I build my idea based on arrayfire-ml, it would probably also trigger something wrong?

Currently, I am rewriting my codes with less but bigger arrays.
I expect less computations for JIT would solve the problem.

But I am still waiting for your good news.
Thanks for helping me.

**Comment by pavanky** [comment]
@BookmanHan I think you are hitting a corner case that arrayfire's JIT compiler isn't catching. I need some time to investigate this.

**Comment by BookmanHan** [comment]
@pavanky 

Thanks, a lot, but I suppose I finally catch the bug.
It is something abnormal happening to `af::tile` in JIT.
When I add `backward.eval();` before `af::tile(backward, 1, ..);` in `View.hpp`, the bug is fixed.

Does this result make sense?
I am not sure if I make it essentially.
Still waiting for detailed explanations if possible, though it is okay temporally.

Thanks, again.

**Comment by pavanky** [comment]
@BookmanHan For element wise operations ArraFire creates an AST internally and only evaluates them we necessary. This ensures a single kernel is created for a lot of operations and decreases the memory used and the number of reads and writes.

For example:

``` c++
a = radu(5); // Creates a buffer
b = a + 1; // Creates a node in the AST
c = a - 1; // Creates another node
d = b * c; // Creates another node
e = sum(d); // sum is not a JIT operation so d.eval() is called internally. A single kernel is lanched to get the value of d.
```

Now the down side is that in some corner cases the AST can get too big and the driver can't handle a large number of inputs. There are some mechanisms in place to avoid issues like this but it looks like you are hitting a corner case that hasn't been considered. Calling `backward.eval()`, you are manually forcing the evaluation and shortening the AST. 

I hope this made sense.

**Comment by WilliamTambellini** [comment]
Hey guys the title is funny but bad: it s not "complication" but "compilation" :)

**Comment by BookmanHan** [comment]
@pavanky It really makes sense to me. Actually, I have triggered this error, again. There is no `af::tile`, this time. Thanks for you explanations. I think you are right about the issue. It is the length of codes that causes the problem. **I extremely force every operation to `eval()`, then this issue is fixed.**

However, current neural networks are bigger and bigger. I have implemented three neural models, two of which hit on this issue. I guess the Seq2Seq(LSTM) model on arrayfire-ml would perform similarly. Before I meet ArrayFire, I think it's fine to run every single operation. But the concept of JIT on GPU is so nice that I feel a little sad about this solution. I really hope ArrayFire could consider the application of bigger neural architecture. 

Thank you, very much.

@WilliamTambellini 
I am very sorry for this mistake. It is really funny but bad.
Thanks for the revision.
